% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/trace_funcs.R
\docType{class}
\name{Trace-class}
\alias{Trace-class}
\alias{Trace}
\title{An S4 class to Hold Genome Track Data}
\description{
An instance of this class is created by calling "bb_makeTrace".  All slots in this object are GRanges objects.  Validation checks will make sure data, peaks, links, and gene models all are bound by the same plot range on the same chromosome of the same genome.  This is different from the standard GRangesList object in that each slot can have it's own metadata columns.  Currently, hg38 and danRer11 are the supported genomes.  Use this class to plot coverage tracks from bulk or single cell ATAC, CHIP, or similar experiments.  Link plotting is available for cicero-style links.
}
\section{Slots}{

\describe{
\item{\code{trace_data}}{A GRanges object with a metadata column for "score" or "coverage" intended to be plotted as a y-variable.  Metadata variables may be included to annotate color and and facets in the final plotting.  This is particularly important for single cell data or when combining bulk tracks from different samples.  Whole sample bigwig files can be converted to GRanges objects using import.bw from rtracklayer.  All track data is trimmed during import but pre-trimming to the approximate range desired will significantly speed up processing.  The plyranges package is recommended for granges manipulations such as filtering by chromosome, adding metadata and binding GRanges objects together.  Bulk tracks from different samples should be pre-normalized before importing.}

\item{\code{peaks}}{A GRanges object containing peaks to plot in a track-style format.}

\item{\code{links}}{A GRanges object with Cicero-style links.}

\item{\code{gene_model}}{The gene model for plotting.  Will be automatically generated by bb_makeTrace.}

\item{\code{plot_range}}{The master GRange for the whole object.  Validity checks and/or constructors ensure all other ranges are contained within.}
}}

