<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="blaseRtools">
<title>Trace:  Representing Track-like Data in R • blaseRtools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Trace:  Representing Track-like Data in R">
<meta property="og:description" content="blaseRtools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9162</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a>
    <a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a>
    <a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a>
    <a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Trace:  Representing Track-like Data in R</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/vignettes/trace.Rmd" class="external-link"><code>vignettes/trace.Rmd</code></a></small>
      <div class="d-none name"><code>trace.Rmd</code></div>
    </div>

    
    
<p>Certain genomic assays are best represented as genomic regions.
Representing a genomic region minimally requires identification of the
relevant sequence as well as a start and end position for the region of
interest. Strand information is often included as well. There are
multiple plain-text and binary formats to store this information. In R,
the implementaion for storing genomic region data is provided by the
GenomicRanges package in the form of a GRanges object. GRanges objects
are flexible and powerful containers for genomic region data. In
addition to these required values, GRanges allows you to store arbitrary
metadata associated with each genomic region or interval. This is very
useful.</p>
<p>This is not a tutorial on how to use GenomicRanges. Please see <a href="https://bioconductor.org/packages/devel/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html" class="external-link">this
instead</a>. Another important resource not discussed in detail here is
the <a href="https://www.bioconductor.org/packages/release/bioc/html/plyranges.html" class="external-link">plyranges
package</a> which provides a dplyr-like interface for manipulating
GRanges objects. Here instead we present the Trace object and associated
functions for visualizing single-cell or bulk genomic region data (ATAC,
CHIP-seq etc.). The Trace object is an extension of GRanges with the
following differences:</p>
<ol style="list-style-type: decimal">
<li>Trace objects have 5 “slots”. Each slot is a GRanges object. The
slots are</li>
</ol>
<ul>
<li>trace_data: The main track you want to show. Usually it will depict
binding or accessibility as a continuous variable across a genomic
region.<br>
</li>
<li>peaks: Defined regions of accessibility or binding represented as
binary values (peak present or not) across a genomic region. Usually
defined by a peak caller such as MACS.</li>
<li>links: Regions of chromatin that are co-accessible. This is a
probabilistic determination based on how often two regions of chromatin
are acessbible in multiple single cells. Does not necessarily represent
a physical interaction between these regions. Not applicable for bulk
assays.</li>
<li>gene_model: The gene model from the reference genome.</li>
<li>plot_range: The overall range of the Trace object.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><p>Validity rules: trace_data, peaks, links and gene model must all
come from the same genome and fit within plot_range.</p></li>
<li><p>Metadata: The metadata columns are not required to be the same in
all of the slots. This is different from the GRangesList
object.</p></li>
</ol>
<p>An example of each of these slots:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/blaserlab/blaseRtools" class="external-link">blaseRtools</a></span><span class="op">)</span></span>
<span><span class="co"># trace_data</span></span>
<span><span class="fu"><a href="../reference/Trace.data.html">Trace.data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 15000 ranges and 7 metadata columns:</span></span>
<span><span class="co">#&gt;           seqnames    ranges strand |     group     count norm.value  coverage</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;factor&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;       [1]     chr6  40525870      * | Erythroid         0          0         0</span></span>
<span><span class="co">#&gt;       [2]     chr6  40526434      * | Erythroid         0          0         0</span></span>
<span><span class="co">#&gt;       [3]     chr6  40525580      * | Erythroid         0          0         0</span></span>
<span><span class="co">#&gt;       [4]     chr6  40526504      * | Erythroid         0          0         0</span></span>
<span><span class="co">#&gt;       [5]     chr6  40542071      * | Erythroid         0          0         0</span></span>
<span><span class="co">#&gt;       ...      ...       ...    ... .       ...       ...        ...       ...</span></span>
<span><span class="co">#&gt;   [14996]     chr6  40520092      * |     Renal         0          0         0</span></span>
<span><span class="co">#&gt;   [14997]     chr6  40525199      * |     Renal         0          0         0</span></span>
<span><span class="co">#&gt;   [14998]     chr6  40528865      * |     Renal         0          0         0</span></span>
<span><span class="co">#&gt;   [14999]     chr6  40521826      * |     Renal         0          0         0</span></span>
<span><span class="co">#&gt;   [15000]     chr6  40529332      * |     Renal         0          0         0</span></span>
<span><span class="co">#&gt;              Assay     assay_group    .group</span></span>
<span><span class="co">#&gt;           &lt;factor&gt;     &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;       [1]    peaks Erythroid_peaks         5</span></span>
<span><span class="co">#&gt;       [2]    peaks Erythroid_peaks         5</span></span>
<span><span class="co">#&gt;       [3]    peaks Erythroid_peaks         5</span></span>
<span><span class="co">#&gt;       [4]    peaks Erythroid_peaks         5</span></span>
<span><span class="co">#&gt;       [5]    peaks Erythroid_peaks         5</span></span>
<span><span class="co">#&gt;       ...      ...             ...       ...</span></span>
<span><span class="co">#&gt;   [14996]    peaks     Renal_peaks         2</span></span>
<span><span class="co">#&gt;   [14997]    peaks     Renal_peaks         2</span></span>
<span><span class="co">#&gt;   [14998]    peaks     Renal_peaks         2</span></span>
<span><span class="co">#&gt;   [14999]    peaks     Renal_peaks         2</span></span>
<span><span class="co">#&gt;   [15000]    peaks     Renal_peaks         2</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from danRer11 genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># peaks</span></span>
<span><span class="fu"><a href="../reference/Trace.peaks.html">Trace.peaks</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 4 ranges and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames            ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr6 40522952-40524640      *</span></span>
<span><span class="co">#&gt;   [2]     chr6 40529989-40531105      *</span></span>
<span><span class="co">#&gt;   [3]     chr6 40531870-40532296      *</span></span>
<span><span class="co">#&gt;   [4]     chr6 40548455-40548860      *</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from danRer11 genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># links</span></span>
<span><span class="fu"><a href="../reference/Trace.links.html">Trace.links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 6 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames            ranges strand |     score     group</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr6 40523796-40530547      * |  0.733665      3109</span></span>
<span><span class="co">#&gt;   [2]     chr6 40523796-40532083      * |  0.447556      3109</span></span>
<span><span class="co">#&gt;   [3]     chr6 40523796-40530547      * |  0.733665      3109</span></span>
<span><span class="co">#&gt;   [4]     chr6 40530547-40532083      * |  0.705522      3109</span></span>
<span><span class="co">#&gt;   [5]     chr6 40523796-40532083      * |  0.447556      3109</span></span>
<span><span class="co">#&gt;   [6]     chr6 40530547-40532083      * |  0.705522      3109</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from danRer11 genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># gene model</span></span>
<span><span class="fu"><a href="../reference/Trace.gene_model.html">Trace.gene_model</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 20 ranges and 6 metadata columns:</span></span>
<span><span class="co">#&gt;        seqnames            ranges strand |      source            type</span></span>
<span><span class="co">#&gt;           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;character&gt;</span></span>
<span><span class="co">#&gt;    [1]     chr6 40523370-40523442      + |        ZFIN  five_prime_UTR</span></span>
<span><span class="co">#&gt;    [2]     chr6 40531976-40531995      + |        ZFIN  five_prime_UTR</span></span>
<span><span class="co">#&gt;    [3]     chr6 40531996-40532104      + |        ZFIN             CDS</span></span>
<span><span class="co">#&gt;    [4]     chr6 40532541-40532740      + |        ZFIN             CDS</span></span>
<span><span class="co">#&gt;    [5]     chr6 40532814-40532880      + |        ZFIN             CDS</span></span>
<span><span class="co">#&gt;    ...      ...               ...    ... .         ...             ...</span></span>
<span><span class="co">#&gt;   [16]     chr6 40541577-40541715      + |        ZFIN             CDS</span></span>
<span><span class="co">#&gt;   [17]     chr6 40542409-40542597      + |        ZFIN             CDS</span></span>
<span><span class="co">#&gt;   [18]     chr6 40543075-40543203      + |        ZFIN             CDS</span></span>
<span><span class="co">#&gt;   [19]     chr6 40543289-40543447      + |        ZFIN             CDS</span></span>
<span><span class="co">#&gt;   [20]     chr6 40543448-40544207      + |        ZFIN three_prime_UTR</span></span>
<span><span class="co">#&gt;         parent_transcript     APPRIS    length   gene_name</span></span>
<span><span class="co">#&gt;               &lt;character&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;    [1] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;    [2] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;    [3] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;    [4] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;    [5] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;    ...                ...        ...       ...         ...</span></span>
<span><span class="co">#&gt;   [16] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;   [17] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;   [18] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;   [19] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;   [20] ENSDART00000033819 principal1      2908      prkcda</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from danRer11 genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># plot range</span></span>
<span><span class="fu"><a href="../reference/Trace.plot_range.html">Trace.plot_range</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 1 range and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames            ranges strand</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr6 40518370-40549207      *</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from danRer11 genome; no seqlengths</span></span></code></pre></div>
<p>The Trace object is suitable for holding such information centered
about a gene of interest. It is as small as possible and can quickly
provide data for very informative plots as we will show.</p>
<div class="section level2">
<h2 id="making-a-trace-object">Making a Trace Object<a class="anchor" aria-label="anchor" href="#making-a-trace-object"></a>
</h2>
<p>Trace objects can be made either from a Signac/Seurat object or from
a GRanges object.</p>
<p>Signac is a package that stores and reads single-cell genomic range
data (accessibility, peaks, etc) to and from a Seurat object. Although
Signac/Seurat objects are powerful and a standard in the community, they
are large, complicated and have lots of hidden slots that are confusing
and error prone for the user. So Trace objects can be used to extract
the required information for visualization in a structured way and to
store the data with a much smaller footprint.</p>
<p>Bulk chromatin analysis outputs data in bigwig files which can be
read into R using the import.bw function from rtracklayer. This will
generate a GRanges object. It will be extremely large and so Trace
objects are much lighter-weight alternative.</p>
<p>You make a Trace object from a Signac/Seurat object like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># note this code will not run</span></span>
<span><span class="va">zf_prkcda_trace</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_makeTrace.html">bb_makeTrace</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">zf</span>,</span>
<span>                             genome <span class="op">=</span> <span class="st">"danRer11"</span>,</span>
<span>                             gene_to_plot <span class="op">=</span> <span class="st">"prkcda"</span>,</span>
<span>                             extend_left <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>                             extend_right <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loaded by blaseRtools</span></span>
<span><span class="va">zf_prkcda_trace</span></span>
<span><span class="co">#&gt; A Trace object from genome danRer11, sequence chr6.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The plot range is set to 40518370-40549207 .</span></span></code></pre></div>
<p>Peaks and Links, if included in the Seurat/signac object will be
transferred to the Trace object.</p>
<p>You make a Trace object from bulk data like so. Filtering the full
GRange will speed up processing considerably.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># note this code will not run</span></span>
<span><span class="va">e4_PRKCD_trace</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_makeTrace.html">bb_makeTrace</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">e4_huvec_GRange_full.pkc_cxcl8</span>, <span class="va">seqnames</span> <span class="op">==</span> <span class="st">"chr3"</span><span class="op">)</span>,</span>
<span>                              gene_to_plot <span class="op">=</span> <span class="st">"PRKCD"</span>,</span>
<span>                              peaks <span class="op">=</span> <span class="va">e4_huvec_peaks.pkc_cxcl8</span>,</span>
<span>                              genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>                              extend_left <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>                              extend_right <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loaded by blaseRtools</span></span>
<span><span class="va">e4_PRKCD_trace</span></span>
<span><span class="co">#&gt; A Trace object from genome hg38, sequence chr3.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The plot range is set to 53151009-53197717 .</span></span></code></pre></div>
<p>When you import a bigwig into a GRanges object, it will not
necessarily have the same group and coverage metadata. It probably will
have only a numeric column called score. If this is the case, the
constructor function will rename it coverage and add a group variable
with the value “bulk”. You can change these options if you bigwig is
different, but leaving them as default (group and coverage) will allow
the plotting functions to work with defaults, so this is what is
recommended.</p>
</div>
<div class="section level2">
<h2 id="plotting-component-tracks">Plotting Component Tracks<a class="anchor" aria-label="anchor" href="#plotting-component-tracks"></a>
</h2>
<p>You will make the plots using a set of functions which I describe
below. Each track is made separately and composed into a single panel
using the patchwork package. Each track is a ggplot object and can be
edited using layers like any other ggplot. Like most plotting functions,
the first argument is always the data object.</p>
<div class="section level3">
<h3 id="ploting-trace-data">Ploting trace data<a class="anchor" aria-label="anchor" href="#ploting-trace-data"></a>
</h3>
<p>You can plot the trace data using bb_plot_trace_data. Without any
other arguments, it looks like this:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span>font_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>For this function, the default y variable is “coverage” and the facet
and color variables are “group”. These are the default because this is
how Signac/Seurat objects hold the data. You set the “group” variable
values by using the Idents function to operate on the Signac object
before making the Trace object. If there are other variables you wish to
display you should add them to the group variable</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Trace.data.html">Trace.data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group_alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">group</span>, <span class="st">"_alt"</span><span class="op">)</span><span class="op">)</span>  </span>
<span><span class="va">zf_prkcda_trace</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Trace.setData.html">Trace.setData</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, gr <span class="op">=</span> <span class="va">dat</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, facet_var <span class="op">=</span> <span class="st">"group_alt"</span>, color_var <span class="op">=</span> <span class="st">"group_alt"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>You can set the color palette:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Trace.data.html">Trace.data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group_alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">group</span>, <span class="st">"_alt"</span><span class="op">)</span><span class="op">)</span>  </span>
<span><span class="va">zf_prkcda_trace</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Trace.setData.html">Trace.setData</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, gr <span class="op">=</span> <span class="va">dat</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">5</span>, name <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="va">alt_pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Erythroid_alt"</span> <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>             <span class="st">"Neutrophil_alt"</span> <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>             <span class="st">"Progenitor 1_alt"</span> <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>             <span class="st">"Progenitor 2_alt"</span> <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>             <span class="st">"Renal_alt"</span> <span class="op">=</span> <span class="va">cols</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span></span>
<span>             <span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, </span>
<span>                   facet_var <span class="op">=</span> <span class="st">"group_alt"</span>, </span>
<span>                   color_var <span class="op">=</span> <span class="st">"group_alt"</span>, </span>
<span>                   pal <span class="op">=</span> <span class="va">alt_pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>You can stack the traces if you want:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, </span>
<span>                   facet_var <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                   color_var <span class="op">=</span> <span class="st">"group_alt"</span>, </span>
<span>                   pal <span class="op">=</span> <span class="va">alt_pal</span>, </span>
<span>                   legend_pos <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>For bulk data, you may have imported the values you want to plot
under another variable name</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Trace.data.html">Trace.data</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 7661 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames            ranges strand |       group  coverage</span></span>
<span><span class="co">#&gt;             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]     chr3 53151013-53151026      * |        bulk   0.20260</span></span>
<span><span class="co">#&gt;      [2]     chr3 53151027-53151029      * |        bulk   0.20197</span></span>
<span><span class="co">#&gt;      [3]     chr3 53151030-53151032      * |        bulk   0.40412</span></span>
<span><span class="co">#&gt;      [4]     chr3 53151033-53151036      * |        bulk   0.60609</span></span>
<span><span class="co">#&gt;      [5]     chr3 53151037-53151069      * |        bulk   0.60799</span></span>
<span><span class="co">#&gt;      ...      ...               ...    ... .         ...       ...</span></span>
<span><span class="co">#&gt;   [7657]     chr3          53197667      * |        bulk   0.17298</span></span>
<span><span class="co">#&gt;   [7658]     chr3 53197668-53197688      * |        bulk   0.08645</span></span>
<span><span class="co">#&gt;   [7659]     chr3 53197689-53197692      * |        bulk   0.08634</span></span>
<span><span class="co">#&gt;   [7660]     chr3 53197693-53197696      * |        bulk   0.08623</span></span>
<span><span class="co">#&gt;   [7661]     chr3 53197697-53197714      * |        bulk   0.08611</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from hg38 genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plotting-the-gene-model">Plotting the Gene Model<a class="anchor" aria-label="anchor" href="#plotting-the-gene-model"></a>
</h3>
<p>This plots the underlying gene model from the reference genome. By
default it picks the APPRIS prinicpal1 transcript for each gene. UTRs
are shown in half the height of coding exons.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Or you can show a specific transcript. There is only 1 transcript for
this gene.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, select_transcript <span class="op">=</span> <span class="st">"ENSDART00000033819"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>Here is the same gene in humans with multiple transcripts:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prinicpal1 transcript</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># an alternative transcript</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span>, select_transcript <span class="op">=</span> <span class="st">"ENST00000651505"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>If you want you can change the fill color of the exon icons:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prinicpal1 transcript</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span>, icon_fill <span class="op">=</span> <span class="st">"orange"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="plotting-peaks">Plotting Peaks<a class="anchor" aria-label="anchor" href="#plotting-peaks"></a>
</h3>
<p>This is relatively straightforward:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>If you want you can change the fill color too:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, fill_color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="plotting-links">Plotting Links<a class="anchor" aria-label="anchor" href="#plotting-links"></a>
</h3>
<p>This function plots bezier curves between the center of linked peaks.
The color in the figure indicates the link score.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-19-1.png" width="700">
In this case, the color scale could be scaled differently to show the
differences more clearly.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, link_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>You can also change the colors.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, </span>
<span>                    link_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.8</span><span class="op">)</span>, link_low_color <span class="op">=</span> <span class="st">"purple"</span>, link_high_color <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-21-1.png" width="700">
If there are link scores outside the range, they will be shown in dark
grey which is a default for ggplot:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, </span>
<span>                    link_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">0.8</span><span class="op">)</span>, </span>
<span>                    link_low_color <span class="op">=</span> <span class="st">"purple"</span>, </span>
<span>                    link_high_color <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>You can also trim those using the cutoff parameter (default is 0, no
cutoff)</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, </span>
<span>                    link_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">0.8</span><span class="op">)</span>, </span>
<span>                    link_low_color <span class="op">=</span> <span class="st">"purple"</span>, </span>
<span>                    link_high_color <span class="op">=</span> <span class="st">"green"</span>, </span>
<span>                    cutoff <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>Note: links that extend outside the plot range are also trimmed.</p>
</div>
<div class="section level3">
<h3 id="plotting-the-axis">Plotting the Axis<a class="anchor" aria-label="anchor" href="#plotting-the-axis"></a>
</h3>
<p>The default shows the plot range in kb, indicating the genome and the
chromosome:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>You can add a custom axis title if you want. You can do this in the
function or as a separate layer.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, xtitle <span class="op">=</span> <span class="st">"My Custom Title"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Another Custom Title"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-25-2.png" width="700"></p>
<p>As we will see in the next section, putting these modifications
within the function calls, rather than as separate layers makes the code
look cleaner.</p>
</div>
</div>
<div class="section level2">
<h2 id="putting-the-plot-together">Putting the Plot Together<a class="anchor" aria-label="anchor" href="#putting-the-plot-together"></a>
</h2>
<p>We use patchwork to compose the tracks into a full plot:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>This needs some help, so we add a plot layout function call:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, link_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.8</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0.01</span><span class="op">)</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>This looks pretty good, but the Coverage plot y axes are a little
cramped at this size:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_links.html">bb_plot_trace_links</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span>, link_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.8</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">zf_prkcda_trace</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0.01</span><span class="op">)</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>This looks better. Note that I have to enclose the ggplot layers in
parentheses to keep from confusing patchwork. Not the end of the
world.</p>
<p>It also works for bulk data:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>This is good, but the facet variable is weird. The “bulk” term is a
placeholder that was introduced when I made the object. I can silence
the facet variable by setting <code>facet_var = NULL</code> to make this
go away.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span>, facet_var <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<p>Or I can put some meaningful information in the facet variable to
serve as a helpful title:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Trace.data.html">Trace.data</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group <span class="op">=</span> <span class="st">"E4-HUVEC"</span><span class="op">)</span></span>
<span><span class="va">e4_PRKCD_trace_alt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Trace.setData.html">Trace.setData</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace</span>, <span class="va">dat</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bb_plot_trace_data.html">bb_plot_trace_data</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace_alt</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_model.html">bb_plot_trace_model</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace_alt</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_peaks.html">bb_plot_trace_peaks</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace_alt</span><span class="op">)</span> <span class="op">/</span> </span>
<span><span class="fu"><a href="../reference/bb_plot_trace_axis.html">bb_plot_trace_axis</a></span><span class="op">(</span><span class="va">e4_PRKCD_trace_alt</span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="trace_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
