<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Infer Cell-Cell Communication — bb_cellchat • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Infer Cell-Cell Communication — bb_cellchat"><meta name="description" content="Use this function to identify ligand/receptor pairs expressed by cell clusters in human, mouse or zebrafish single cell data.  A CellChat object is generated which can be used to visualize these connections using bb_cellchat_heatmap or other tools from package CellChat."><meta property="og:description" content="Use this function to identify ligand/receptor pairs expressed by cell clusters in human, mouse or zebrafish single cell data.  A CellChat object is generated which can be used to visualize these connections using bb_cellchat_heatmap or other tools from package CellChat."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9188</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Infer Cell-Cell Communication</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/cellchat.R" class="external-link"><code>R/cellchat.R</code></a></small>
      <div class="d-none name"><code>bb_cellchat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Use this function to identify ligand/receptor pairs expressed by cell clusters in human, mouse or zebrafish single cell data.  A CellChat object is generated which can be used to visualize these connections using bb_cellchat_heatmap or other tools from package CellChat.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_cellchat</span><span class="op">(</span></span>
<span>  <span class="va">cds</span>,</span>
<span>  <span class="va">group_var</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"mouse"</span>, <span class="st">"zebrafish"</span><span class="op">)</span>,</span>
<span>  min_cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  prob_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"triMean"</span>, <span class="st">"truncatedMean"</span>, <span class="st">"median"</span><span class="op">)</span>,</span>
<span>  prob_trim <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  project <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pop_size_arg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ask <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cds">cds<a class="anchor" aria-label="anchor" href="#arg-cds"></a></dt>
<dd><p>The cell data set object.  It should usually be pre-filtered to conatin a single biological sample.</p></dd>


<dt id="arg-group-var">group_var<a class="anchor" aria-label="anchor" href="#arg-group-var"></a></dt>
<dd><p>The cell metadata column identifying cell groups for cell-cell communication inference.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>Number of cores for the analysis, Default: 12</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>Species for the assay, Default: c("human", "mouse", "zebrafish")</p></dd>


<dt id="arg-min-cells">min_cells<a class="anchor" aria-label="anchor" href="#arg-min-cells"></a></dt>
<dd><p>Cell clusters smaller than this value will be ignored., Default: 10</p></dd>


<dt id="arg-prob-type">prob_type<a class="anchor" aria-label="anchor" href="#arg-prob-type"></a></dt>
<dd><p>Methods for computing the average gene expression per cell group. By default = "triMean", producing fewer but stronger interactions; When setting ‘type = "truncatedMean"', a value should be assigned to ’trim', producing more interactions, Default: c("triMean", "truncatedMean", "median")</p></dd>


<dt id="arg-prob-trim">prob_trim<a class="anchor" aria-label="anchor" href="#arg-prob-trim"></a></dt>
<dd><p>the fraction (0 to 0.25) of observations to be trimmed from each end of x before the mean is computed if using truncatedMean, Default: NULL</p></dd>


<dt id="arg-project">project<a class="anchor" aria-label="anchor" href="#arg-project"></a></dt>
<dd><p>Whether or not to smooth gene expression, Default: TRUE</p></dd>


<dt id="arg-pop-size-arg">pop_size_arg<a class="anchor" aria-label="anchor" href="#arg-pop-size-arg"></a></dt>
<dd><p>Whether consider the proportion of cells in each group across all sequenced cells. Set population.size = FALSE if analyzing sorting-enriched single cells, to remove the potential artifact of population size. Set population.size = TRUE if analyzing unsorted single-cell transcriptomes, with the reason that abundant cell populations tend to send collectively stronger signals than the rare cell populations., Default: TRUE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A CellChat object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>see github::sqjin/CellChat</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/monocle3/man/normalized_counts.html" class="external-link">normalized_counts</a></code>
<code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">mutate-joins</a></code>,<code><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></code>
<code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></code>
<code><a href="https://rdrr.io/pkg/CellChat/man/createCellChat.html" class="external-link">createCellChat</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/CellChatDB.human.html" class="external-link">CellChatDB.human</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/CellChatDB.mouse.html" class="external-link">CellChatDB.mouse</a></code>,<code>character(0)</code>,<code><a href="https://rdrr.io/pkg/CellChat/man/subsetData.html" class="external-link">subsetData</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/identifyOverExpressedGenes.html" class="external-link">identifyOverExpressedGenes</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/identifyOverExpressedInteractions.html" class="external-link">identifyOverExpressedInteractions</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/projectData.html" class="external-link">projectData</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/computeCommunProb.html" class="external-link">computeCommunProb</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/filterCommunication.html" class="external-link">filterCommunication</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/computeCommunProbPathway.html" class="external-link">computeCommunProbPathway</a></code>,<code><a href="https://rdrr.io/pkg/CellChat/man/aggregateNet.html" class="external-link">aggregateNet</a></code>
<code>character(0)</code>
<code><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

