<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Supply a cds subset and aggregation values and get a heatmap.  Plot the return value using cowplot::plot_grid().  This function wraps several complicated functions from ComplexHeatmap and tries to apply default values for the most common use case:  plotting top markers from a cds with cells grouped arbitrarily (usually by cluster of some type).  This function provides the option to aggregate by genes as well in order to plot gene modules.  For cell and gene aggregation, provide a name (in the form of a string) of the corresponding metadata column and the aggregation will be performed.  There are many aesthetic parameters for this function and even more available for the internal ComplexHeatmap::Heatmap() function.  The best way to adjust parameters not provided here is to scroll the popup box that appears when you type ComplexHeatmap::Heatmap() in RStudio and pass them in via the ellipsis.  The default is to put genes in columns and cells in rows.  You can flip this behavior by setting flip_axis = TRUE.  This will require adjustment of some aesthetic parameters.  Complex annotations (beyond labeling cherry-picked genes) are not currently supported."><title>Make a Heatmap of Aggregated Gene Expressionf from a CDS — bb_cds_heatmap • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Make a Heatmap of Aggregated Gene Expressionf from a CDS — bb_cds_heatmap"><meta property="og:description" content="Supply a cds subset and aggregation values and get a heatmap.  Plot the return value using cowplot::plot_grid().  This function wraps several complicated functions from ComplexHeatmap and tries to apply default values for the most common use case:  plotting top markers from a cds with cells grouped arbitrarily (usually by cluster of some type).  This function provides the option to aggregate by genes as well in order to plot gene modules.  For cell and gene aggregation, provide a name (in the form of a string) of the corresponding metadata column and the aggregation will be performed.  There are many aesthetic parameters for this function and even more available for the internal ComplexHeatmap::Heatmap() function.  The best way to adjust parameters not provided here is to scroll the popup box that appears when you type ComplexHeatmap::Heatmap() in RStudio and pass them in via the ellipsis.  The default is to put genes in columns and cells in rows.  You can flip this behavior by setting flip_axis = TRUE.  This will require adjustment of some aesthetic parameters.  Complex annotations (beyond labeling cherry-picked genes) are not currently supported."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9143</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a>
    <a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Make a Heatmap of Aggregated Gene Expressionf from a CDS</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/bb_cds_heatmap.R" class="external-link"><code>R/bb_cds_heatmap.R</code></a></small>
      <div class="d-none name"><code>bb_cds_heatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Supply a cds subset and aggregation values and get a heatmap.  Plot the return value using cowplot::plot_grid().  This function wraps several complicated functions from ComplexHeatmap and tries to apply default values for the most common use case:  plotting top markers from a cds with cells grouped arbitrarily (usually by cluster of some type).  This function provides the option to aggregate by genes as well in order to plot gene modules.  For cell and gene aggregation, provide a name (in the form of a string) of the corresponding metadata column and the aggregation will be performed.  There are many aesthetic parameters for this function and even more available for the internal ComplexHeatmap::Heatmap() function.  The best way to adjust parameters not provided here is to scroll the popup box that appears when you type ComplexHeatmap::Heatmap() in RStudio and pass them in via the ellipsis.  The default is to put genes in columns and cells in rows.  You can flip this behavior by setting flip_axis = TRUE.  This will require adjustment of some aesthetic parameters.  Complex annotations (beyond labeling cherry-picked genes) are not currently supported.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_cds_heatmap</span><span class="op">(</span></span>
<span>  <span class="va">cds_subset</span>,</span>
<span>  cellmeta_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rowmeta_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  heatmap_highlights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  three_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue4"</span>, <span class="st">"ivory"</span>, <span class="st">"red3"</span><span class="op">)</span>,</span>
<span>  flip_axis <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>title_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fontface <span class="op">=</span> <span class="st">"plain"</span>, fontsize <span class="op">=</span> <span class="fl">9</span><span class="op">)</span>,</span>
<span>    grid_width <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">0.14</span>, <span class="st">"in"</span><span class="op">)</span>, labels_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  row_dend_width <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">5</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  column_dend_height <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">5</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  column_dend_side <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  row_names_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">9</span><span class="op">)</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  row_dend_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>lwd <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  column_dend_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>lwd <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  row_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  column_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  padding <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  labels_rot <span class="op">=</span> <span class="fl">45</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>cds_subset</dt>
<dd><p>The subset of cells and genes you want to plot as a heatmap.  Best approach is to pipe the cds through filter_cds() and into this function.</p></dd>


<dt>cellmeta_col</dt>
<dd><p>The name of a cell metadata column to aggregate cells by; one of cellmeta_col and rowmeta_col must not be NULL, Default: NULL</p></dd>


<dt>rowmeta_col</dt>
<dd><p>The name of a row metadata column to aggregate cells by; one of cellmeta_col and rowmeta_col must not be NULL, Default: NULL</p></dd>


<dt>heatmap_highlights</dt>
<dd><p>A vector of gene names to highlight using anno_mark(), Default: NULL</p></dd>


<dt>three_colors</dt>
<dd><p>A vector of colors for the main color scale, Default: c("blue4", "ivory", "red3")</p></dd>


<dt>flip_axis</dt>
<dd><p>Logical; whether to plot genes as rows (TRUE) or columns (FALSE), Default: FALSE</p></dd>


<dt>name</dt>
<dd><p>Name of the main color scale, Default: NULL</p></dd>


<dt>heatmap_legend_param</dt>
<dd><p>Graphical parameters for the main heatmap legend, Default: list(title_gp = gpar(fontface = "plain", fontsize = 9), grid_width = unit(0.14,
"in"), labels_gp = gpar(fontsize = 8))</p></dd>


<dt>row_dend_width</dt>
<dd><p>Row dendrogram width, Default: unit(5, "mm")</p></dd>


<dt>column_dend_height</dt>
<dd><p>Column dendrogram height, Default: unit(5, "mm")</p></dd>


<dt>column_dend_side</dt>
<dd><p>Side on which to plot the column dendrogram, Default: 'bottom'</p></dd>


<dt>show_row_names</dt>
<dd><p>Logical; whether or not to show rownames, Default: T</p></dd>


<dt>row_names_gp</dt>
<dd><p>Graphical parameters for the row names, Default: gpar(fontsize = 9)</p></dd>


<dt>show_column_names</dt>
<dd><p>Logical; whether or not to show column names, Default: F</p></dd>


<dt>row_dend_gp</dt>
<dd><p>Graphical parameters for the row dendrogram, Default: gpar(lwd = 0.5)</p></dd>


<dt>column_dend_gp</dt>
<dd><p>Graphical parameters for teh column dendrogram, Default: gpar(lwd = 0.5)</p></dd>


<dt>row_title</dt>
<dd><p>Row title text, Default: NULL</p></dd>


<dt>column_title</dt>
<dd><p>Column title text, Default: NULL</p></dd>


<dt>padding</dt>
<dd><p>Padding between gene names on the heatmap highlights, Default: 1.5</p></dd>


<dt>labels_rot</dt>
<dd><p>Rotation of the heatmap highlight labels, Default: 45</p></dd>


<dt>...</dt>
<dd><p>Optional arguments to pass to ComplexHeatmap::Heatmap()</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A complex heatmap in the form of a gtree.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

