<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Make a Plot of Gene Expression in UMAP Form — bb_gene_umap • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Make a Plot of Gene Expression in UMAP Form — bb_gene_umap"><meta name="description" content='Takes in a Seurat or cell_dat_set object, extracts UMAP dimensions and gene expression values.  For Seurat, default assay is "RNA"; can be changed to if necessary.  For cell_data_set, the assay parameter does nothing; the function extracts log and size-factor normalized counts which are similar but not identical to the Seurat "RNA" assay.  If a vector of genes is supplied to gene_or_genes, a faceted plot will be generated.  If a dataframe is supplied, an aggregated plot will be generated with a facet for each gene group.  The dataframe must be of 2 colums: the first containing feature ids and the second containing grouping information.  This is best generated using bb_rowmeta.'><meta property="og:description" content='Takes in a Seurat or cell_dat_set object, extracts UMAP dimensions and gene expression values.  For Seurat, default assay is "RNA"; can be changed to if necessary.  For cell_data_set, the assay parameter does nothing; the function extracts log and size-factor normalized counts which are similar but not identical to the Seurat "RNA" assay.  If a vector of genes is supplied to gene_or_genes, a faceted plot will be generated.  If a dataframe is supplied, an aggregated plot will be generated with a facet for each gene group.  The dataframe must be of 2 colums: the first containing feature ids and the second containing grouping information.  This is best generated using bb_rowmeta.'></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9184</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Make a Plot of Gene Expression in UMAP Form</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/bb_gene_umap.R" class="external-link"><code>R/bb_gene_umap.R</code></a></small>
      <div class="d-none name"><code>bb_gene_umap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Takes in a Seurat or cell_dat_set object, extracts UMAP dimensions and gene expression values.  For Seurat, default assay is "RNA"; can be changed to if necessary.  For cell_data_set, the assay parameter does nothing; the function extracts log and size-factor normalized counts which are similar but not identical to the Seurat "RNA" assay.  If a vector of genes is supplied to gene_or_genes, a faceted plot will be generated.  If a dataframe is supplied, an aggregated plot will be generated with a facet for each gene group.  The dataframe must be of 2 colums: the first containing feature ids and the second containing grouping information.  This is best generated using bb_rowmeta.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_gene_umap</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  <span class="va">gene_or_genes</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cell_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color_legend_title <span class="op">=</span> <span class="st">"Expression"</span>,</span>
<span>  max_expr_val <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alt_dim_x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alt_dim_y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rasterize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  raster_dpi <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  cds <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>A Seurat or cell_data_set object.</p></dd>


<dt id="arg-gene-or-genes">gene_or_genes<a class="anchor" aria-label="anchor" href="#arg-gene-or-genes"></a></dt>
<dd><p>Individual gene or genes or aggregated genes to plot.  Supply a character string for a single gene, a vector for multiple genes or a dataframe for aggregated genes.  See description.</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>For Seurat objects only:  the gene expression assay to get expression data from, Default: 'RNA'</p></dd>


<dt id="arg-order">order<a class="anchor" aria-label="anchor" href="#arg-order"></a></dt>
<dd><p>Whether or not to order cells by gene expression.  When ordered, non-expressing cells are plotted first, i.e. on the bottom.  Caution:  when many cells are overplotted it may lead to a misleading presentation.  Generally bb_genebubbles is a better way to present, Default: TRUE</p></dd>


<dt id="arg-cell-size">cell_size<a class="anchor" aria-label="anchor" href="#arg-cell-size"></a></dt>
<dd><p>Size of the points, Default: 1</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>Transparency of the points, Default: 1</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>Specify the number of columns if faceting, Default: NULL</p></dd>


<dt id="arg-plot-title">plot_title<a class="anchor" aria-label="anchor" href="#arg-plot-title"></a></dt>
<dd><p>Optional title for the plot, Default: NULL</p></dd>


<dt id="arg-color-legend-title">color_legend_title<a class="anchor" aria-label="anchor" href="#arg-color-legend-title"></a></dt>
<dd><p>Option to change the color scale title, Default: 'Expression'</p></dd>


<dt id="arg-max-expr-val">max_expr_val<a class="anchor" aria-label="anchor" href="#arg-max-expr-val"></a></dt>
<dd><p>Maximum expression value to cap the color scale, Default: NULL</p></dd>


<dt id="arg-alt-dim-x">alt_dim_x<a class="anchor" aria-label="anchor" href="#arg-alt-dim-x"></a></dt>
<dd><p>Alternate/reference dimensions to plot by.</p></dd>


<dt id="arg-alt-dim-y">alt_dim_y<a class="anchor" aria-label="anchor" href="#arg-alt-dim-y"></a></dt>
<dd><p>Alternate/reference dimensions to plot by.</p></dd>


<dt id="arg-rasterize">rasterize<a class="anchor" aria-label="anchor" href="#arg-rasterize"></a></dt>
<dd><p>Whether to render the graphical layer as a raster image.  Default is FALSE.</p></dd>


<dt id="arg-raster-dpi">raster_dpi<a class="anchor" aria-label="anchor" href="#arg-raster-dpi"></a></dt>
<dd><p>If rasterize then this is the DPI used.  Default = 300.</p></dd>


<dt id="arg-cds">cds<a class="anchor" aria-label="anchor" href="#arg-cds"></a></dt>
<dd><p>Provided for backward compatibility.  If a value is supplied a warning will be emitted., Default: NULL</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/monocle3/man/normalized_counts.html" class="external-link">normalized_counts</a></code>
<code><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></code>
<code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></code>
<code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">mutate-joins</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></code>
<code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_d</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

