<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Make an Ape Genome Object — bb_make_ape_genomic • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Make an Ape Genome Object — bb_make_ape_genomic"><meta name="description" content='This function takes either a gene name or sequence coordinates and returns an ape object with DNA sequence from the the selected genome reference.  You can choose from hg38 and GRCz11.  Features are generated from ensembl GFF files.  Features overlapping the query range (gene or sequence) are returned as a GRanges object and as features within the Ape object.  The features included can optionally be filtered using the "include_type" argument to the function. Using the "additional_granges" argument, you can provide additional features not present in the standard gene model which will be added to the Ape object GRanges slot and to the features slot.'><meta property="og:description" content='This function takes either a gene name or sequence coordinates and returns an ape object with DNA sequence from the the selected genome reference.  You can choose from hg38 and GRCz11.  Features are generated from ensembl GFF files.  Features overlapping the query range (gene or sequence) are returned as a GRanges object and as features within the Ape object.  The features included can optionally be filtered using the "include_type" argument to the function. Using the "additional_granges" argument, you can provide additional features not present in the standard gene model which will be added to the Ape object GRanges slot and to the features slot.'></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9175</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Make an Ape Genome Object</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/ape_make.R" class="external-link"><code>R/ape_make.R</code></a></small>
      <div class="d-none name"><code>bb_make_ape_genomic.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function takes either a gene name or sequence coordinates and returns an ape object with DNA sequence from the the selected genome reference.  You can choose from hg38 and GRCz11.  Features are generated from ensembl GFF files.  Features overlapping the query range (gene or sequence) are returned as a GRanges object and as features within the Ape object.  The features included can optionally be filtered using the "include_type" argument to the function. Using the "additional_granges" argument, you can provide additional features not present in the standard gene model which will be added to the Ape object GRanges slot and to the features slot.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_make_ape_genomic</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  genome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg38"</span>, <span class="st">"GRCz11"</span><span class="op">)</span>,</span>
<span>  extend_left <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  extend_right <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  include_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ncRNA_gene"</span>, <span class="st">"rRNA"</span>, <span class="st">"exon"</span>, <span class="st">"pseudogene"</span>, <span class="st">"pseudogenic_transcript"</span>,</span>
<span>    <span class="st">"ncRNA"</span>, <span class="st">"gene"</span>, <span class="st">"CDS"</span>, <span class="st">"lnc_RNA"</span>, <span class="st">"mRNA"</span>, <span class="st">"three_prime_UTR"</span>, <span class="st">"five_prime_UTR"</span>,</span>
<span>    <span class="st">"unconfirmed_transcript"</span>, <span class="st">"scRNA"</span>, <span class="st">"C_gene_segment"</span>, <span class="st">"D_gene_segment"</span>,</span>
<span>    <span class="st">"J_gene_segment"</span>, <span class="st">"V_gene_segment"</span>, <span class="st">"miRNA"</span>, <span class="st">"tRNA"</span>, <span class="st">"snRNA"</span>, <span class="st">"snoRNA"</span>,</span>
<span>    <span class="st">"lincRNA_gene"</span>, <span class="st">"lncRNA_gene"</span>, <span class="st">"unconfirmed_transcript"</span><span class="op">)</span>,</span>
<span>  additional_granges <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-query">query<a class="anchor" aria-label="anchor" href="#arg-query"></a></dt>
<dd><p>Either a valid gene name or a named numeric vector of genome coordinates.  This vector should be of the form:  c(chr = 1, start = 1000, end = 2000).  The vector must be numeric an must have those names.  The chromosome number will be converted to "chr1" etc internally.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>The genome to pull from, Default: c("hg38", "GRCz11")</p></dd>


<dt id="arg-extend-left">extend_left<a class="anchor" aria-label="anchor" href="#arg-extend-left"></a></dt>
<dd><p>Number of bases to extend the query to the left or "upstream" relative to the + strand.</p></dd>


<dt id="arg-extend-right">extend_right<a class="anchor" aria-label="anchor" href="#arg-extend-right"></a></dt>
<dd><p>Number of bases to extend the query to the right or "downstream" relative to the + strand.</p></dd>


<dt id="arg-include-type">include_type<a class="anchor" aria-label="anchor" href="#arg-include-type"></a></dt>
<dd><p>The type of features to include from the standard gene model. Default: c("ncRNA_gene", "rRNA", "exon", "pseudogene", "pseudogenic_transcript",
"ncRNA", "gene", "CDS", "lnc_RNA", "mRNA", "three_prime_UTR",
"five_prime_UTR", "unconfirmed_transcript", "scRNA", "C_gene_segment",
"D_gene_segment", "J_gene_segment", "V_gene_segment", "miRNA",
"tRNA", "snRNA", "snoRNA", "lincRNA_gene", "lncRNA_gene",
"unconfirmed_transcript")</p></dd>


<dt id="arg-additional-granges">additional_granges<a class="anchor" aria-label="anchor" href="#arg-additional-granges"></a></dt>
<dd><p>A GRanges object with features to add to the Ape Object.  Coordinates should all be relative to the reference, <em>NOT</em> the sequence extracted for the ape file.  The Granges object can be constructed with the following syntax:  GenomicRanges::makeGRangesFromDataFrame(data.frame(seqname = "chr6", start = 40523370, end = 40523380, strand = "+", type = "addl_feature", gene_name = "prkcda", label = "feature1"), keep.extra.columns = T).  The gene_name argument here is optional.  If you have defined features based on the extracted sequence, (i.e. relative to position 1 in the ORIGIN section of the Ape object), the best option is to use the feature setting function FEATURES(instance_of_Ape) &lt;- GRanges_Object.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An APE object</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

