<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate a CDS from Bulk RNA-seq data — bb_pseudocells • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a CDS from Bulk RNA-seq data — bb_pseudocells"><meta name="description" content="This function allows you to simulate single cell data from bulk RNA-seq data.  It requires a TPM count matrix.  The rationale is that TPM quantifies transcript counts per million reads, so you can think of this like 1 million UMI counts from a scRNA-seq experiment distributed in a certain way across the transcriptome.  This function samples n_pseudocells  with transcripts_per_pseudocell from this distribution.  Then it creates a cell data set based on a matrix of these samples.
Importantly, this function cannot accurately identify transcriptional heterogenity within the bulk data.  The sampling effect may reveal some potential heterogeneity but there is no method here for determining whether this is due to randomness or heterogeneity within the data.
The intended use of this function is to project a pseudocell cds onto an actual single cell data set.  This can be used to help identify regions of UMAP space that are shared between the pseudocells and the real cells.
Note:  The matrix rownames and row metadata for the returned cds will contain only the rownames from the input tpm_matrix.  So these should share the same namespace as the single cell cds that the pseudocell data will be projected onto."><meta property="og:description" content="This function allows you to simulate single cell data from bulk RNA-seq data.  It requires a TPM count matrix.  The rationale is that TPM quantifies transcript counts per million reads, so you can think of this like 1 million UMI counts from a scRNA-seq experiment distributed in a certain way across the transcriptome.  This function samples n_pseudocells  with transcripts_per_pseudocell from this distribution.  Then it creates a cell data set based on a matrix of these samples.
Importantly, this function cannot accurately identify transcriptional heterogenity within the bulk data.  The sampling effect may reveal some potential heterogeneity but there is no method here for determining whether this is due to randomness or heterogeneity within the data.
The intended use of this function is to project a pseudocell cds onto an actual single cell data set.  This can be used to help identify regions of UMAP space that are shared between the pseudocells and the real cells.
Note:  The matrix rownames and row metadata for the returned cds will contain only the rownames from the input tpm_matrix.  So these should share the same namespace as the single cell cds that the pseudocell data will be projected onto."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9188</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Generate a CDS from Bulk RNA-seq data</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/pseudocells.R" class="external-link"><code>R/pseudocells.R</code></a></small>
      <div class="d-none name"><code>bb_pseudocells.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function allows you to simulate single cell data from bulk RNA-seq data.  It requires a TPM count matrix.  The rationale is that TPM quantifies transcript counts per million reads, so you can think of this like 1 million UMI counts from a scRNA-seq experiment distributed in a certain way across the transcriptome.  This function samples <em>n_pseudocells</em>  with <em>transcripts_per_pseudocell</em> from this distribution.  Then it creates a cell data set based on a matrix of these samples.</p>
<p>Importantly, this function cannot accurately identify transcriptional heterogenity within the bulk data.  The sampling effect may reveal some potential heterogeneity but there is no method here for determining whether this is due to randomness or heterogeneity within the data.</p>
<p>The intended use of this function is to project a pseudocell cds onto an actual single cell data set.  This can be used to help identify regions of UMAP space that are shared between the pseudocells and the real cells.</p>
<p>Note:  The matrix rownames and row metadata for the returned cds will contain only the rownames from the input tpm_matrix.  So these should share the same namespace as the single cell cds that the pseudocell data will be projected onto.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_pseudocells</span><span class="op">(</span></span>
<span>  <span class="va">tpm_matrix</span>,</span>
<span>  <span class="va">n_pseudocells</span>,</span>
<span>  <span class="va">transcripts_per_pseudocell</span>,</span>
<span>  remove_genes <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tpm-matrix">tpm_matrix<a class="anchor" aria-label="anchor" href="#arg-tpm-matrix"></a></dt>
<dd><p>A matrix of TPM counts from a bulk RNA experiment.  A cds will be generated for each column in this matrix and the result combined.</p></dd>


<dt id="arg-n-pseudocells">n_pseudocells<a class="anchor" aria-label="anchor" href="#arg-n-pseudocells"></a></dt>
<dd><p>The number of pseudocells to create.  Should be length 1 or to specify a uniqe n_pseudocells for each dataset, a vector of the same length as the number of columns in tpm_matrix.  This value will be recycled if necessary to match the number of columns in tpm_mtx.</p></dd>


<dt id="arg-transcripts-per-pseudocell">transcripts_per_pseudocell<a class="anchor" aria-label="anchor" href="#arg-transcripts-per-pseudocell"></a></dt>
<dd><p>The number of transcripts to sample for each pseudocell.  Should be similar to the median number of UMI in the single cell data the pseudocells will be projected onto.  Should be length 1 or to specify a uniqe n_pseudocells for each dataset, a vector of the same length as the number of columns in tpm_matrix.  This value will be recycled if necessary to match the number of columns in tpm_mtx.</p></dd>


<dt id="arg-remove-genes">remove_genes<a class="anchor" aria-label="anchor" href="#arg-remove-genes"></a></dt>
<dd><p>A vector of genes to remove before sampling.  Should be the same or similar to the genes removed from the single cell data the pseudocells will be projected onto, Default: NULL</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A cell data set</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></code>, <code><a href="https://purrr.tidyverse.org/reference/reduce.html" class="external-link">reduce</a></code>, <code><a href="https://purrr.tidyverse.org/reference/map2.html" class="external-link">map2</a></code>, <code><a href="https://purrr.tidyverse.org/reference/pmap.html" class="external-link">pmap</a></code>
<code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></code>
<code><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></code>
<code><a href="https://generics.r-lib.org/reference/components.html" class="external-link">components</a></code>
<code><a href="https://rdrr.io/pkg/monocle3/man/new_cell_data_set.html" class="external-link">new_cell_data_set</a></code>, <code><a href="https://rdrr.io/pkg/monocle3/man/combine_cds.html" class="external-link">combine_cds</a></code>, <code><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html" class="external-link">preprocess_cds</a></code>, <code><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html" class="external-link">reduce_dimension</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span></span>
<span class="r-in"><span> <span class="co">#EXAMPLE1</span></span></span>
<span class="r-in"><span> <span class="op">}</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

