<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>A function to generate a UMAP with colors mapped to colData variables — bb_var_umap • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A function to generate a UMAP with colors mapped to colData variables — bb_var_umap"><meta name="description" content="A function to generate a UMAP with colors mapped to colData variables"><meta property="og:description" content="A function to generate a UMAP with colors mapped to colData variables"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9188</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>A function to generate a UMAP with colors mapped to colData variables</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/bb_var_umap.R" class="external-link"><code>R/bb_var_umap.R</code></a></small>
      <div class="d-none name"><code>bb_var_umap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A function to generate a UMAP with colors mapped to colData variables</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_var_umap</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  <span class="va">var</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  value_to_highlight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  foreground_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  legend_pos <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>  cell_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  alt_stroke_color <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  palette <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alt_dim_x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alt_dim_y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite_labels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  group_label_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  alt_label_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">21</span>,</span>
<span>  nbin <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  facet_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_equally <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rasterize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  raster_dpi <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  show_trajectory_graph <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trajectory_graph_color <span class="op">=</span> <span class="st">"grey28"</span>,</span>
<span>  trajectory_graph_segment_size <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  label_root_node <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pseudotime_dim <span class="op">=</span> <span class="va">var</span>,</span>
<span>  label_principal_points <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  graph_label_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  cds <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  outline_cluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outline_color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  outline_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  outline_type <span class="op">=</span> <span class="st">"solid"</span>,</span>
<span>  outline_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  man_text_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  text_geom <span class="op">=</span> <span class="st">"text"</span>,</span>
<span>  minimum_segment_length <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  hexify <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_hexbins <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>A Seurat or cell data set object</p></dd>


<dt id="arg-var">var<a class="anchor" aria-label="anchor" href="#arg-var"></a></dt>
<dd><p>The variable to map colors to.  Special exceptions are "density", "local_n" and "log_local_n" which calculate the 2 d kernel density estimate or binned cell counts and maps to color scale.</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>The gene expression assay to draw reduced dimensions from.  Default is "RNA".  Does not do anything with cell_data_set objects.</p></dd>


<dt id="arg-value-to-highlight">value_to_highlight<a class="anchor" aria-label="anchor" href="#arg-value-to-highlight"></a></dt>
<dd><p>Option to highlight a single value</p></dd>


<dt id="arg-foreground-alpha">foreground_alpha<a class="anchor" aria-label="anchor" href="#arg-foreground-alpha"></a></dt>
<dd><p>Alpha value for foreground points</p></dd>


<dt id="arg-legend-pos">legend_pos<a class="anchor" aria-label="anchor" href="#arg-legend-pos"></a></dt>
<dd><p>Legend position</p></dd>


<dt id="arg-cell-size">cell_size<a class="anchor" aria-label="anchor" href="#arg-cell-size"></a></dt>
<dd><p>Cell point size</p></dd>


<dt id="arg-alt-stroke-color">alt_stroke_color<a class="anchor" aria-label="anchor" href="#arg-alt-stroke-color"></a></dt>
<dd><p>Alternative color for the data point stroke</p></dd>


<dt id="arg-legend-title">legend_title<a class="anchor" aria-label="anchor" href="#arg-legend-title"></a></dt>
<dd><p>Title for the legend</p></dd>


<dt id="arg-plot-title">plot_title<a class="anchor" aria-label="anchor" href="#arg-plot-title"></a></dt>
<dd><p>Main title for the plot</p></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>Color palette to use.  "Rcolorbrewer", "Viridis" are builtin options.  Otherwise provide manual values.</p></dd>


<dt id="arg-alt-dim-x">alt_dim_x<a class="anchor" aria-label="anchor" href="#arg-alt-dim-x"></a></dt>
<dd><p>Alternate/reference dimensions to plot by.</p></dd>


<dt id="arg-alt-dim-y">alt_dim_y<a class="anchor" aria-label="anchor" href="#arg-alt-dim-y"></a></dt>
<dd><p>Alternate/reference dimensions to plot by.</p></dd>


<dt id="arg-overwrite-labels">overwrite_labels<a class="anchor" aria-label="anchor" href="#arg-overwrite-labels"></a></dt>
<dd><p>Whether to overwrite the variable value labels</p></dd>


<dt id="arg-group-label-size">group_label_size<a class="anchor" aria-label="anchor" href="#arg-group-label-size"></a></dt>
<dd><p>Size of the overwritten labels</p></dd>


<dt id="arg-alt-label-col">alt_label_col<a class="anchor" aria-label="anchor" href="#arg-alt-label-col"></a></dt>
<dd><p>Alternate column to label cells by</p></dd>


<dt id="arg-shape">shape<a class="anchor" aria-label="anchor" href="#arg-shape"></a></dt>
<dd><p>Shape for data points</p></dd>


<dt id="arg-nbin">nbin<a class="anchor" aria-label="anchor" href="#arg-nbin"></a></dt>
<dd><p>Number of bins if using var %in% c("density". "local_n", "log_local_n")</p></dd>


<dt id="arg-facet-by">facet_by<a class="anchor" aria-label="anchor" href="#arg-facet-by"></a></dt>
<dd><p>Variable or variables to facet by.</p></dd>


<dt id="arg-sample-equally">sample_equally<a class="anchor" aria-label="anchor" href="#arg-sample-equally"></a></dt>
<dd><p>Whether or not you should downsample to the same number of cells in each plot.  Default is FALSE or no.</p></dd>


<dt id="arg-rasterize">rasterize<a class="anchor" aria-label="anchor" href="#arg-rasterize"></a></dt>
<dd><p>Whether to render the graphical layer as a raster image.  Default is FALSE.</p></dd>


<dt id="arg-raster-dpi">raster_dpi<a class="anchor" aria-label="anchor" href="#arg-raster-dpi"></a></dt>
<dd><p>If rasterize then this is the DPI used.  Default = 300.</p></dd>


<dt id="arg-show-trajectory-graph">show_trajectory_graph<a class="anchor" aria-label="anchor" href="#arg-show-trajectory-graph"></a></dt>
<dd><p>Whether to render the principal graph for the
trajectory. Requires that learn_graph() has been called on cds.</p></dd>


<dt id="arg-trajectory-graph-color">trajectory_graph_color<a class="anchor" aria-label="anchor" href="#arg-trajectory-graph-color"></a></dt>
<dd><p>The color to be used for plotting the
trajectory graph.</p></dd>


<dt id="arg-trajectory-graph-segment-size">trajectory_graph_segment_size<a class="anchor" aria-label="anchor" href="#arg-trajectory-graph-segment-size"></a></dt>
<dd><p>The size of the line segments used for
plotting the trajectory graph.</p></dd>


<dt id="arg-label-root-node">label_root_node<a class="anchor" aria-label="anchor" href="#arg-label-root-node"></a></dt>
<dd><p>Logical; whether to label the root node for the selected pseudotime trajectory.  The function will requires that a valid pseudotime column be identified, usually as the value of the "var" argument in the form of "pseudotime_cluster_value".  If you wish to use var to color the cells in some other way, the pseudotime_dim argument needs to be supplied with the correct pseudotime dimension to pick the root node from.</p></dd>


<dt id="arg-pseudotime-dim">pseudotime_dim<a class="anchor" aria-label="anchor" href="#arg-pseudotime-dim"></a></dt>
<dd><p>An alternative column to pick the pseudoetime root node from, if not supplied to var.</p></dd>


<dt id="arg-label-principal-points">label_principal_points<a class="anchor" aria-label="anchor" href="#arg-label-principal-points"></a></dt>
<dd><p>Logical indicating whether to label roots,
leaves, and branch points with principal point names. This is useful for
order_cells and choose_graph_segments in non-interactive mode.</p></dd>


<dt id="arg-graph-label-size">graph_label_size<a class="anchor" aria-label="anchor" href="#arg-graph-label-size"></a></dt>
<dd><p>How large to make the branch, root, and leaf labels.</p></dd>


<dt id="arg-cds">cds<a class="anchor" aria-label="anchor" href="#arg-cds"></a></dt>
<dd><p>Provided for backward compatibility with prior versions.  If a value is supplied, a warning will be emitted and the value will be transferred to the obj argument, Default: NULL</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional params for facetting.</p></dd>


<dt id="arg-man-text-df">man_text_df<a class="anchor" aria-label="anchor" href="#arg-man-text-df"></a></dt>
<dd><p>A data frame in the form of text_x = numeric_vector, text_y = numeric_vector, label = character_vector for manually placing text labels.</p></dd>


<dt id="arg-minimum-segment-length">minimum_segment_length<a class="anchor" aria-label="anchor" href="#arg-minimum-segment-length"></a></dt>
<dd><p>Minimum length of a line to draw from label to centroid.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a ggplot</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

