<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A function to generate a UMAP with colors mapped to colData variables"><title>A function to generate a UMAP with colors mapped to colData variables — bb_var_umap • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A function to generate a UMAP with colors mapped to colData variables — bb_var_umap"><meta property="og:description" content="A function to generate a UMAP with colors mapped to colData variables"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9148</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a>
    <a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a>
    <a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a>
    <a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>A function to generate a UMAP with colors mapped to colData variables</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/bb_var_umap.R" class="external-link"><code>R/bb_var_umap.R</code></a></small>
      <div class="d-none name"><code>bb_var_umap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A function to generate a UMAP with colors mapped to colData variables</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_var_umap</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  <span class="va">var</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  value_to_highlight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  foreground_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  legend_pos <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>  cell_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  alt_stroke_color <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  palette <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alt_dim_x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alt_dim_y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite_labels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  group_label_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  alt_label_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">21</span>,</span>
<span>  nbin <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  facet_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_equally <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rasterize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  raster_dpi <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  show_trajectory_graph <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trajectory_graph_color <span class="op">=</span> <span class="st">"grey28"</span>,</span>
<span>  trajectory_graph_segment_size <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  label_root_node <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pseudotime_dim <span class="op">=</span> <span class="va">var</span>,</span>
<span>  label_principal_points <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  graph_label_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  cds <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  outline_cluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outline_color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  outline_size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  outline_type <span class="op">=</span> <span class="st">"solid"</span>,</span>
<span>  outline_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  man_text_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  text_geom <span class="op">=</span> <span class="st">"text"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>obj</dt>
<dd><p>A Seurat or cell data set object</p></dd>


<dt>var</dt>
<dd><p>The variable to map colors to.  Special exceptions are "density", "local_n" and "log_local_n" which calculate the 2 d kernel density estimate or binned cell counts and maps to color scale.</p></dd>


<dt>assay</dt>
<dd><p>The gene expression assay to draw reduced dimensions from.  Default is "RNA".  Does not do anything with cell_data_set objects.</p></dd>


<dt>value_to_highlight</dt>
<dd><p>Option to highlight a single value</p></dd>


<dt>foreground_alpha</dt>
<dd><p>Alpha value for foreground points</p></dd>


<dt>legend_pos</dt>
<dd><p>Legend position</p></dd>


<dt>cell_size</dt>
<dd><p>Cell point size</p></dd>


<dt>alt_stroke_color</dt>
<dd><p>Alternative color for the data point stroke</p></dd>


<dt>legend_title</dt>
<dd><p>Title for the legend</p></dd>


<dt>plot_title</dt>
<dd><p>Main title for the plot</p></dd>


<dt>palette</dt>
<dd><p>Color palette to use.  "Rcolorbrewer", "Viridis" are builtin options.  Otherwise provide manual values.</p></dd>


<dt>alt_dim_x</dt>
<dd><p>Alternate/reference dimensions to plot by.</p></dd>


<dt>alt_dim_y</dt>
<dd><p>Alternate/reference dimensions to plot by.</p></dd>


<dt>overwrite_labels</dt>
<dd><p>Whether to overwrite the variable value labels</p></dd>


<dt>group_label_size</dt>
<dd><p>Size of the overwritten labels</p></dd>


<dt>alt_label_col</dt>
<dd><p>Alternate column to label cells by</p></dd>


<dt>shape</dt>
<dd><p>Shape for data points</p></dd>


<dt>nbin</dt>
<dd><p>Number of bins if using var %in% c("density". "local_n", "log_local_n")</p></dd>


<dt>facet_by</dt>
<dd><p>Variable or variables to facet by.</p></dd>


<dt>sample_equally</dt>
<dd><p>Whether or not you should downsample to the same number of cells in each plot.  Default is FALSE or no.</p></dd>


<dt>rasterize</dt>
<dd><p>Whether to render the graphical layer as a raster image.  Default is FALSE.</p></dd>


<dt>raster_dpi</dt>
<dd><p>If rasterize then this is the DPI used.  Default = 300.</p></dd>


<dt>show_trajectory_graph</dt>
<dd><p>Whether to render the principal graph for the
trajectory. Requires that learn_graph() has been called on cds.</p></dd>


<dt>trajectory_graph_color</dt>
<dd><p>The color to be used for plotting the
trajectory graph.</p></dd>


<dt>trajectory_graph_segment_size</dt>
<dd><p>The size of the line segments used for
plotting the trajectory graph.</p></dd>


<dt>label_root_node</dt>
<dd><p>Logical; whether to label the root node for the selected pseudotime trajectory.  The function will requires that a valid pseudotime column be identified, usually as the value of the "var" argument in the form of "pseudotime_cluster_value".  If you wish to use var to color the cells in some other way, the pseudotime_dim argument needs to be supplied with the correct pseudotime dimension to pick the root node from.</p></dd>


<dt>pseudotime_dim</dt>
<dd><p>An alternative column to pick the pseudoetime root node from, if not supplied to var.</p></dd>


<dt>label_principal_points</dt>
<dd><p>Logical indicating whether to label roots,
leaves, and branch points with principal point names. This is useful for
order_cells and choose_graph_segments in non-interactive mode.</p></dd>


<dt>graph_label_size</dt>
<dd><p>How large to make the branch, root, and leaf labels.</p></dd>


<dt>cds</dt>
<dd><p>Provided for backward compatibility with prior versions.  If a value is supplied, a warning will be emitted and the value will be transferred to the obj argument, Default: NULL</p></dd>


<dt>...</dt>
<dd><p>Additional params for facetting.</p></dd>


<dt>man_text_df</dt>
<dd><p>A data frame in the form of text_x = numeric_vector, text_y = numeric_vector, label = character_vector for manually placing text labels.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a ggplot</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

