<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9189</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="Ape-class.html">Ape-class</a></code> <code><a href="Ape-class.html">Ape</a></code>

        </dt>
        <dd>An S4 class to hold genebank/APE file data.</dd>
      </dl><dl><dt>

          <code><a href="Ape.DNA.html">Ape.DNA()</a></code>

        </dt>
        <dd>Get the DNASringSet Slot from an Ape Object</dd>
      </dl><dl><dt>

          <code><a href="Ape.fasta.html">Ape.fasta()</a></code>

        </dt>
        <dd>Save an Ape Instance as a Fasta File</dd>
      </dl><dl><dt>

          <code><a href="Ape.fimo.html">Ape.fimo()</a></code>

        </dt>
        <dd>Run FIMO on Selected Ape Object Features</dd>
      </dl><dl><dt>

          <code><a href="Ape.granges.html">Ape.granges()</a></code>

        </dt>
        <dd>Get the GRanges Slot from an Ape Object</dd>
      </dl><dl><dt>

          <code><a href="Ape.save.html">Ape.save()</a></code>

        </dt>
        <dd>Save an Ape Instance as a Genebank Format File</dd>
      </dl><dl><dt>

          <code><a href="Ape.setFeatures.html">Ape.setFeatures()</a></code>

        </dt>
        <dd>Set the FEATURES Slot of a GRanges Object</dd>
      </dl><dl><dt>

          <code><a href="COMMENTS.html">COMMENTS()</a></code>

        </dt>
        <dd>Get the Comments Slot from an Ape Object</dd>
      </dl><dl><dt>

          <code><a href="FEATURES.html">FEATURES()</a></code>

        </dt>
        <dd>Get the Features Slot from an Ape Object</dd>
      </dl><dl><dt>

          <code><a href="Image-class.html">show(<i>&lt;Image_&gt;</i>)</a></code> <code><a href="Image-class.html">file_path()</a></code> <code><a href="Image-class.html">`file_path&lt;-`()</a></code> <code><a href="Image-class.html">species()</a></code> <code><a href="Image-class.html">`species&lt;-`()</a></code> <code><a href="Image-class.html">stage()</a></code> <code><a href="Image-class.html">`stage&lt;-`()</a></code> <code><a href="Image-class.html">genetics()</a></code> <code><a href="Image-class.html">`genetics&lt;-`()</a></code> <code><a href="Image-class.html">treatment()</a></code> <code><a href="Image-class.html">`treatment&lt;-`()</a></code> <code><a href="Image-class.html">microscope()</a></code> <code><a href="Image-class.html">`microscope&lt;-`()</a></code> <code><a href="Image-class.html">mag()</a></code> <code><a href="Image-class.html">`mag&lt;-`()</a></code> <code><a href="Image-class.html">filter()</a></code> <code><a href="Image-class.html">`filter&lt;-`()</a></code> <code><a href="Image-class.html">use()</a></code> <code><a href="Image-class.html">`use&lt;-`()</a></code> <code><a href="Image-class.html">note()</a></code> <code><a href="Image-class.html">`note&lt;-`()</a></code>

        </dt>
        <dd>An S4 Class for Holding Image Metadata</dd>
      </dl><dl><dt>

          <code><a href="ImageCatalog-class.html">show(<i>&lt;ImageCatalog&gt;</i>)</a></code> <code><a href="ImageCatalog-class.html">`[`(<i>&lt;ImageCatalog&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="ImageCatalog-class.html">`[[`(<i>&lt;ImageCatalog&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="ImageCatalog-class.html">`$`(<i>&lt;ImageCatalog&gt;</i>)</a></code> <code><a href="ImageCatalog-class.html">ImageCatalog.as_tibble()</a></code> <code><a href="ImageCatalog-class.html">ImageCatalog.write()</a></code> <code><a href="ImageCatalog-class.html">ImageCatalog.add()</a></code> <code><a href="ImageCatalog-class.html">ImageCatalog.delete()</a></code>

        </dt>
        <dd>An S4 Image Catalog</dd>
      </dl><dl><dt>

          <code><a href="LOCUS.html">LOCUS()</a></code>

        </dt>
        <dd>Get the Locus Slot from an Ape Object</dd>
      </dl><dl><dt>

          <code><a href="LearnGraph.html">LearnGraph()</a></code>

        </dt>
        <dd>Run <code>link[monocle3]{learn_graph}</code> on a <code><a href="https://satijalab.org/seurat/reference/Seurat-package.html" class="external-link">Seurat</a></code> object</dd>
      </dl><dl><dt>

          <code><a href="SeuratWrappers-package.html">blaseRtools</a></code> <code><a href="SeuratWrappers-package.html">blaseRtools-package</a></code> <code><a href="SeuratWrappers-package.html">SeuratWrappers-package</a></code>

        </dt>
        <dd>blaseRtools: R Tools for Blaser Lab Data Analysis</dd>
      </dl><dl><dt>

          <code><a href="SummarizedHeatmap.html">SummarizedHeatmap()</a></code>

        </dt>
        <dd>An S4 Class for Holding Heatmap Data</dd>
      </dl><dl><dt>

          <code><a href="Trace-class.html">Trace-class</a></code> <code><a href="Trace-class.html">Trace</a></code>

        </dt>
        <dd>An S4 class to Hold Genome Track Data</dd>
      </dl><dl><dt>

          <code><a href="Trace.data.html">Trace.data()</a></code>

        </dt>
        <dd>Get the Trace Data Slot from a Trace Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.gene_model.html">Trace.gene_model()</a></code>

        </dt>
        <dd>Get the gene_model Slot from a Trace Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.links.html">Trace.links()</a></code>

        </dt>
        <dd>Get the Links Slot from a Trace Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.peaks.html">Trace.peaks()</a></code>

        </dt>
        <dd>Get the peaks Slot from a Trace Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.plot_range.html">Trace.plot_range()</a></code>

        </dt>
        <dd>Get the plot_range Slot from a Trace Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.setData.html">Trace.setData()</a></code>

        </dt>
        <dd>Set the Trace Data Slot of a GRanges Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.setLinks.html">Trace.setLinks()</a></code>

        </dt>
        <dd>Set the Links Slot of a GRanges Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.setRange.html">Trace.setRange()</a></code>

        </dt>
        <dd>Set the Plot Range Slot of a GRanges Object</dd>
      </dl><dl><dt>

          <code><a href="Trace.setpeaks.html">Trace.setpeaks()</a></code>

        </dt>
        <dd>Set the peaks Slot of a GRanges Object</dd>
      </dl><dl><dt>

          <code><a href="add_cds_factor_columns.html">add_cds_factor_columns()</a></code>

        </dt>
        <dd>Add predefined sample-level cell metadata to cell data sets while importing</dd>
      </dl><dl><dt>

          <code><a href="aggregate.Matrix.html">aggregate(<i>&lt;Matrix&gt;</i>)</a></code>

        </dt>
        <dd>Compute summary statistics of a Matrix</dd>
      </dl><dl><dt>

          <code><a href="as.Seurat.extras.html">as.Seurat(<i>&lt;cell_data_set&gt;</i>)</a></code>

        </dt>
        <dd>Convert single cell experiment to Seurat</dd>
      </dl><dl><dt>

          <code><a href="as.cell_data_set.html">as.cell_data_set()</a></code>

        </dt>
        <dd>Convert objects to Monocle3 <code>cell_data_set</code> objects</dd>
      </dl><dl><dt>

          <code><a href="bb_aggregate.html">bb_aggregate()</a></code>

        </dt>
        <dd>Aggregate Single Cell Gene Expression</dd>
      </dl><dl><dt>

          <code><a href="bb_align.html">bb_align()</a></code>

        </dt>
        <dd>Align a CDS object according to a metadata variable</dd>
      </dl><dl><dt>

          <code><a href="bb_annotate_npc.html">bb_annotate_npc()</a></code>

        </dt>
        <dd>Annotate a Plot using NPC Coordinates</dd>
      </dl><dl><dt>

          <code><a href="bb_blind_images.html">bb_blind_images()</a></code>

        </dt>
        <dd>Make a Copy of Image Files and Rename With File Hashes in Blinded Folder</dd>
      </dl><dl><dt>

          <code><a href="bb_buff_granges.html">bb_buff_granges()</a></code>

        </dt>
        <dd>Clean Up Your GRanges Object</dd>
      </dl><dl><dt>

          <code><a href="bb_cds_anno.html">bb_cds_anno()</a></code>

        </dt>
        <dd>Annotate a CDS by Label Transfer</dd>
      </dl><dl><dt>

          <code><a href="bb_cds_heatmap.html">bb_cds_heatmap()</a></code>

        </dt>
        <dd>Make a Heatmap of Aggregated Gene Expressionf from a CDS</dd>
      </dl><dl><dt>

          <code><a href="bb_cellchat.html">bb_cellchat()</a></code>

        </dt>
        <dd>Infer Cell-Cell Communication</dd>
      </dl><dl><dt>

          <code><a href="bb_cellchat_heatmap.html">bb_cellchat_heatmap()</a></code>

        </dt>
        <dd>Make a Complex Heatmap From a CellChat Object With Sensible Defaults</dd>
      </dl><dl><dt>

          <code><a href="bb_cellmeta.html">bb_cellmeta()</a></code>

        </dt>
        <dd>Get Cell Metadata</dd>
      </dl><dl><dt>

          <code><a href="bb_cite_umap.html">bb_cite_umap()</a></code>

        </dt>
        <dd>Plot a UMAP Showing Cite-Seq Antibody Binding</dd>
      </dl><dl><dt>

          <code><a href="bb_cluster_representation.html">bb_cluster_representation()</a></code>

        </dt>
        <dd>Cluster Representation By Fisher Exact Test Per Cell</dd>
      </dl><dl><dt>

          <code><a href="bb_cluster_representation2.html">bb_cluster_representation2()</a></code>

        </dt>
        <dd>Cluster Representation By Regression Per Sample</dd>
      </dl><dl><dt>

          <code><a href="bb_daseq.html">bb_daseq()</a></code>

        </dt>
        <dd>Calculate Differential Abundance Using DAseq</dd>
      </dl><dl><dt>

          <code><a href="bb_doubletfinder.html">bb_doubletfinder()</a></code>

        </dt>
        <dd>Use doubletfinder to model and mark doublets</dd>
      </dl><dl><dt>

          <code><a href="bb_extract_msig.html">bb_extract_msig()</a></code>

        </dt>
        <dd>Extract MSIGDB Gene Sets</dd>
      </dl><dl><dt>

          <code><a href="bb_fix_file_path.html">bb_fix_file_path()</a></code>

        </dt>
        <dd>Convert windows filepath to linux-compatible.</dd>
      </dl><dl><dt>

          <code><a href="bb_fragment_replacement.html">bb_fragment_replacement()</a></code>

        </dt>
        <dd>Replace The Path to a Fragment File In a Signac Object</dd>
      </dl><dl><dt>

          <code><a href="bb_gene_dotplot.html">bb_gene_dotplot()</a></code>

        </dt>
        <dd>Make a dotplot of gene expression by cell population</dd>
      </dl><dl><dt>

          <code><a href="bb_gene_modules.html">bb_gene_modules()</a></code>

        </dt>
        <dd>A function to generate gene modules and add them to the Gene Metadata</dd>
      </dl><dl><dt>

          <code><a href="bb_gene_pseudotime.html">bb_gene_pseudotime()</a></code>

        </dt>
        <dd>Plot expression of a gene or genes in pseudotime.</dd>
      </dl><dl><dt>

          <code><a href="bb_gene_umap.html">bb_gene_umap()</a></code>

        </dt>
        <dd>Make a Plot of Gene Expression in UMAP Form</dd>
      </dl><dl><dt>

          <code><a href="bb_gene_violinplot.html">bb_gene_violinplot()</a></code>

        </dt>
        <dd>Make a plot of gene expression in UMAP form</dd>
      </dl><dl><dt>

          <code><a href="bb_genebubbles.html">bb_genebubbles()</a></code>

        </dt>
        <dd>Create a Gene Bubble/Dot Plot</dd>
      </dl><dl><dt>

          <code><a href="bb_goenrichment.html">bb_goenrichment()</a></code>

        </dt>
        <dd>Go Term Enrichment</dd>
      </dl><dl><dt>

          <code><a href="bb_goscatter.html">bb_goscatter()</a></code>

        </dt>
        <dd>Make a scatter plot of GO term associations</dd>
      </dl><dl><dt>

          <code><a href="bb_gosummary.html">bb_gosummary()</a></code>

        </dt>
        <dd>A function to reduce go terms by semantic similarity</dd>
      </dl><dl><dt>

          <code><a href="bb_import_bw.html">bb_import_bw()</a></code>

        </dt>
        <dd>Import a Big Wig File</dd>
      </dl><dl><dt>

          <code><a href="bb_import_seacr_peaks.html">bb_import_seacr_peaks()</a></code> <code><a href="bb_import_seacr_peaks.html">bb_import_macs_narrowpeaks()</a></code>

        </dt>
        <dd>Import Peaks from SEACR</dd>
      </dl><dl><dt>

          <code><a href="bb_load_tenx_h5.html">bb_load_tenx_h5()</a></code>

        </dt>
        <dd>Load a 10X Genomics H5 File and Return a CDS</dd>
      </dl><dl><dt>

          <code><a href="bb_load_tenx_targz.html">bb_load_tenx_targz()</a></code>

        </dt>
        <dd>Load 10X Data Into CDS</dd>
      </dl><dl><dt>

          <code><a href="bb_loupeR.html">bb_loupeR()</a></code>

        </dt>
        <dd>Create a Loupe File from a Cell Data Set</dd>
      </dl><dl><dt>

          <code><a href="bb_makeTrace.html">bb_makeTrace()</a></code>

        </dt>
        <dd>Construct a Trace object from a Signac Object or Granges object</dd>
      </dl><dl><dt>

          <code><a href="bb_make_ape_genomic.html">bb_make_ape_genomic()</a></code>

        </dt>
        <dd>Make an Ape Genome Object</dd>
      </dl><dl><dt>

          <code><a href="bb_make_ape_transcript.html">bb_make_ape_transcript()</a></code>

        </dt>
        <dd>Make an Ape Transcriptome Object</dd>
      </dl><dl><dt>

          <code><a href="bb_merge_narrowpeaks.html">bb_merge_narrowpeaks()</a></code>

        </dt>
        <dd>A Function To Merge Replicate Narrow Peaks GRanges</dd>
      </dl><dl><dt>

          <code><a href="bb_metafeature.html">bb_metafeature()</a></code>

        </dt>
        <dd>A Function to Generate Data For Making MetaPlots</dd>
      </dl><dl><dt>

          <code><a href="bb_monocle_.html">bb_monocle_()</a></code> <code><a href="bb_monocle_.html">bb_monocle_anno()</a></code> <code><a href="bb_monocle_.html">bb_monocle_project()</a></code>

        </dt>
        <dd>Annotate Single cell Data Using Monocle3</dd>
      </dl><dl><dt>

          <code><a href="bb_monocle_regression.html">bb_monocle_regression()</a></code>

        </dt>
        <dd>A function to perform regression on single cell data.</dd>
      </dl><dl><dt>

          <code><a href="bb_monocle_regression_better.html">bb_monocle_regression_better()</a></code>

        </dt>
        <dd>An Improved Function to Perform Regression on Single Cell Data.</dd>
      </dl><dl><dt>

          <code><a href="bb_parseape.html">bb_parseape()</a></code>

        </dt>
        <dd>Parse a Genebank File and Construct an APE Object</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_genes_in_pseudotime.html">bb_plot_genes_in_pseudotime()</a></code>

        </dt>
        <dd>Plots expression for one or more genes as a function of pseudotime</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_heatmap_colData.html">bb_plot_heatmap_colData()</a></code>

        </dt>
        <dd>Plot SummarizedHeatmap colData</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_heatmap_colDendro.html">bb_plot_heatmap_colDendro()</a></code>

        </dt>
        <dd>Plot a Heatmap Column Dendrogram</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_heatmap_colHighlight.html">bb_plot_heatmap_colHighlight()</a></code>

        </dt>
        <dd>Plot a Column Highlight</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_heatmap_main.html">bb_plot_heatmap_main()</a></code>

        </dt>
        <dd>Plot the Body of Heatmap</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_heatmap_rowData.html">bb_plot_heatmap_rowData()</a></code>

        </dt>
        <dd>Plot a SummarizedHeatmap rowData</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_heatmap_rowDendro.html">bb_plot_heatmap_rowDendro()</a></code>

        </dt>
        <dd>Plot a Heatmap Row Dendrogram</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_heatmap_rowHighlight.html">bb_plot_heatmap_rowHighlight()</a></code>

        </dt>
        <dd>Plot a Row Highlight</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_rowData_col.html">bb_plot_rowData_col()</a></code>

        </dt>
        <dd>A helper function to generate a data frame in the proper form for aggregate expression plotting with bb_gene_umap.</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_trace_axis.html">bb_plot_trace_axis()</a></code>

        </dt>
        <dd>Plot The X Axis From A Trace Object</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_trace_data.html">bb_plot_trace_data()</a></code>

        </dt>
        <dd>Plot The Trace Data From A Trace Object</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_trace_links.html">bb_plot_trace_links()</a></code>

        </dt>
        <dd>Plot The Link Data From A Trace Object</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_trace_model.html">bb_plot_trace_model()</a></code>

        </dt>
        <dd>Plot The Gene Model From A Trace Object</dd>
      </dl><dl><dt>

          <code><a href="bb_plot_trace_peaks.html">bb_plot_trace_peaks()</a></code>

        </dt>
        <dd>Plot The peak Data From A Trace Object</dd>
      </dl><dl><dt>

          <code><a href="bb_plotfootprint.html">bb_plotfootprint()</a></code>

        </dt>
        <dd>Plot motif footprinting results</dd>
      </dl><dl><dt>

          <code><a href="bb_print_full_stats.html">bb_print_full_stats()</a></code>

        </dt>
        <dd>Print out a stats report</dd>
      </dl><dl><dt>

          <code><a href="bb_promoter_overlap.html">bb_promoter_overlap()</a></code>

        </dt>
        <dd>Calculate The Overlap Between Peaks and Promoters</dd>
      </dl><dl><dt>

          <code><a href="bb_pseudobulk_mf.html">bb_pseudobulk_mf()</a></code>

        </dt>
        <dd>Run Multifactor Pseudobulk Analysis using Deseq2</dd>
      </dl><dl><dt>

          <code><a href="bb_pseudocells.html">bb_pseudocells()</a></code>

        </dt>
        <dd>Generate a CDS from Bulk RNA-seq data</dd>
      </dl><dl><dt>

          <code><a href="bb_pseudotime.html">bb_pseudotime()</a></code>

        </dt>
        <dd>Learn Graph and Calculate Pseudotime</dd>
      </dl><dl><dt>

          <code><a href="bb_qc.html">bb_qc()</a></code>

        </dt>
        <dd>A function to run qc tests on cds objects.</dd>
      </dl><dl><dt>

          <code><a href="bb_read_bam.html">bb_read_bam()</a></code>

        </dt>
        <dd>Read Bam Files</dd>
      </dl><dl><dt>

          <code><a href="bb_read_narrowpeak.html">bb_read_narrowpeak()</a></code>

        </dt>
        <dd>A Function to Read In Narrow Peaks Files</dd>
      </dl><dl><dt>

          <code><a href="bb_rejoin.html">bb_rejoin()</a></code>

        </dt>
        <dd>Rejoin qc and doubletfinder data to a cds object</dd>
      </dl><dl><dt>

          <code><a href="bb_remove_dupes.html">bb_remove_dupes()</a></code>

        </dt>
        <dd>Remove rows that have duplicates in a given column</dd>
      </dl><dl><dt>

          <code><a href="bb_rowmeta.html">bb_rowmeta()</a></code>

        </dt>
        <dd>Get Feature/Gene Metadata</dd>
      </dl><dl><dt>

          <code><a href="bb_seurat_anno.html">bb_seurat_anno()</a></code>

        </dt>
        <dd>A function to generate automated cell labelings with Seurat</dd>
      </dl><dl><dt>

          <code><a href="bb_split_atac.html">bb_split_atac()</a></code>

        </dt>
        <dd>Split Out Peaks Data</dd>
      </dl><dl><dt>

          <code><a href="bb_split_citeseq.html">bb_split_citeseq()</a></code>

        </dt>
        <dd>Split Antibody Capture Data into Alt Experiment</dd>
      </dl><dl><dt>

          <code><a href="bb_tbl_to_coldata.html">bb_tbl_to_coldata()</a></code>

        </dt>
        <dd>A Function To Add Tibble Columns To Cell Metadata</dd>
      </dl><dl><dt>

          <code><a href="bb_tbl_to_matrix.html">bb_tbl_to_matrix()</a></code>

        </dt>
        <dd>Convert a wide-form tibble a matrix</dd>
      </dl><dl><dt>

          <code><a href="bb_tbl_to_rowdata.html">bb_tbl_to_rowdata()</a></code>

        </dt>
        <dd>A Function To Add Tibble Columns To Feature Metadata</dd>
      </dl><dl><dt>

          <code><a href="bb_triplecluster.html">bb_triplecluster()</a></code>

        </dt>
        <dd>A function to generate clusters from scRNA-seq data</dd>
      </dl><dl><dt>

          <code><a href="bb_unblind_images.html">bb_unblind_images()</a></code>

        </dt>
        <dd>Rejoin Blinded Scores to Original File Names</dd>
      </dl><dl><dt>

          <code><a href="bb_var_umap.html">bb_var_umap()</a></code>

        </dt>
        <dd>A function to generate a UMAP with colors mapped to colData variables</dd>
      </dl><dl><dt>

          <code><a href="dMcast.html">dMcast()</a></code>

        </dt>
        <dd>Casts or pivots a long <code>data frame</code> into a wide sparse matrix</dd>
      </dl><dl><dt>

          <code><a href="data_median_se.html">data_median_se()</a></code>

        </dt>
        <dd>Generate median +/- se stat object for jitter ggplot</dd>
      </dl><dl><dt>

          <code><a href="data_summary_mean_sd.html">data_summary_mean_sd()</a></code>

        </dt>
        <dd>Generate mean +/- sd stat object for jitter ggplot</dd>
      </dl><dl><dt>

          <code><a href="data_summary_mean_se.html">data_summary_mean_se()</a></code>

        </dt>
        <dd>Generate mean +/- se stat object for jitter ggplot</dd>
      </dl><dl><dt>

          <code><a href="data_summary_median_iqr.html">data_summary_median_iqr()</a></code>

        </dt>
        <dd>Generate median +/- iqr stat object for jitter ggplot</dd>
      </dl><dl><dt>

          <code><a href="data_summary_median_mad.html">data_summary_median_mad()</a></code>

        </dt>
        <dd>Generate median +/- mad stat object for jitter ggplot</dd>
      </dl><dl><dt>

          <code><a href="filter_cds.html">filter_cds()</a></code>

        </dt>
        <dd>Filter a CDS</dd>
      </dl><dl><dt>

          <code><a href="geom_split_violin.html">geom_split_violin()</a></code>

        </dt>
        <dd>geom_split_violin</dd>
      </dl><dl><dt>

          <code><a href="granges_to_features.html">granges_to_features()</a></code>

        </dt>
        <dd>Format a GRanges Object as a Character Vector For Inclusion in Genebank File</dd>
      </dl><dl><dt>

          <code><a href="grapes-notin-grapes.html">`%notin%`</a></code>

        </dt>
        <dd>Operators</dd>
      </dl><dl><dt>

          <code><a href="merge.Matrix.html">merge.Matrix()</a></code> <code><a href="merge.Matrix.html">join.Matrix()</a></code>

        </dt>
        <dd>Merges two Matrices or matrix-like objects</dd>
      </dl><dl><dt>

          <code><a href="normalize_batch.html">normalize_batch()</a></code>

        </dt>
        <dd>Noramalize a Data Table by Group and Batch</dd>
      </dl><dl><dt>

          <code><a href="rBind.fill.html">rBind.fill()</a></code>

        </dt>
        <dd>Combine matrixes by row, fill in missing columns</dd>
      </dl><dl><dt>

          <code><a href="se.html">se()</a></code>

        </dt>
        <dd>Calculate standard error of the mean</dd>
      </dl><dl><dt>

          <code><a href="show-Ape-method.html">show(<i>&lt;Ape&gt;</i>)</a></code>

        </dt>
        <dd>Show an Ape Object</dd>
      </dl><dl><dt>

          <code><a href="show-Trace-method.html">show(<i>&lt;Trace&gt;</i>)</a></code>

        </dt>
        <dd>Show a Trace Object</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

